### ISSUE TITLE
[Bug] asynchronous event handling is principally impossible

### ISSUE THREAD

**Issue by rozek** [bigskysoftware/_hyperscript#362]
Well,

I tried to evaluate the "async transparency" of \_hyperscript within event handlers...and failed. To my current knowledge of JS, it is principally impossible to implement it anyway:

* JS event handling expects event handlers to run synchronously - only then `preventDefault` and `stop(Immediate)Propagation` are guaranteed to work at any time
* "yielding" from an event handler (to use a term from JS "generators"), e.g., because of an actually async command, does not prevent JS from propagating the event (unless `stop(Immediate)Propagation` was called _before_)
* the `halt` command at the end of a \_hyperscript event handler is therefore completely useless if the code executed before was actually async - but because of "async transparency" the user/programmer probably does not even know what kind of code he/she is running

I have made a small [live example](https://svelte.dev/repl/c5695220b3d346f4ae420d4698d307ce) to illustrate the problem.

According to my research, in JS, async event handlers (which allow `preventDefault` and `stop(Immediate)Propagation` at any time, even after async calls) are principally impossible

And even if this restriction can be circumvented somehow (e.g. by defining propagation/default handling prior to execution) one still has to implement a separate event queuing mechanism in order to keep the original order of incoming events intact - regardless of the duration of any async calls (this is what you tried, but according to my example, the order of events is not preserved)

Not preserving the order of events and/or letting several event handlers run in parallel opens the door for nasty **race conditions** making programming much more complicated than before. Particularly, as people may not recognize this danger just by simply "reading" \_hyperscript code: "...It is very easy to read, making it obvious what a script is doing..." turns out not to be true any more...

**Comment by rozek** [comment]
Some (rough) ideas for potential solutions for the mentioned problem:

* forget about `halt` and let bubbling/default-handling be *configured* separately - perhaps depending on event filters (which must then be synchronous, though)
* alternatively: whenever a scripted event handler turns out to be actually asynchronous, automatically *inhibit* any further event propagation and default handling - this sounds quite strict but may be the only safe way...

Additionally,

* move any events that are to be handled by a script into a separate queue (similar to the one you have already). Within that queue, make sure that the original event order is preserved, i.e., no new event will be handled before handling the previous one has finished
* perhaps introduce a `pass` command (instead of `halt`) that explicitly passes an event to any outer elements (you may have to implement this "custom bubbling" yourself and, perhaps, restrict it to triggering scripted event handlers only)
* perhaps explicitly allow for default handling from within a scripted event handler (e.g., as part of the new `pass` command) - this could probably be implemented by cloning the original event (but now with option `bubbles` set to `false`) and dispatching it again to the original `event.target` (disadvantage: the original `event.target` will get the event twice) Setting `bubbles` to `false` will keep the event away from any outer elements and prevent infinite loops

Perhaps, these ideas may help you saving \_hyperscript as a whole - otherwise, the goals behind \_hyperscript (e.g., "async transparency" and "It is very easy to read, making it obvious what a script is doing") will no longer be achieved...

**Comment by rozek** [comment]
In order to illustrate the potential fixes mentioned above, I've made a small live example

* [this one](https://svelte.dev/repl/4843f5381b4f4704ae04ed08ec9ac5b4) demonstrates how to auto-disable bubbling and default handling when an event handler turns out to be asynchronous

**Comment by 1cg** [comment]
halt doesn't necessarily stop the event handler, you can halt just the event bubbling or prevent default and continue execution:

https://hyperscript.org/commands/halt/

so you can do this at the start of the script and then have whatever async behavior you want afterwards

if you have a conditional situation, where event propagation is dependent on an async result, the easiest thing would be to trigger a separate event based on that result and listen for that on the parent:

```
  on click
     fetch /whatever as json
     if the result's whatever is true
       trigger it-was-true
```

I can imagine detecting when an event handler doesn't explicitly return from execution (which implies async behavior occurred), cancelling the event, and then, when it does finally return, cloning the event and re-dispatching it on the parent.

**Comment by rozek** [comment]
Well,

I have no doubts that people finally find ways to adjust their scripts in order to achieve the intended result

It is that "async transparency" implies that the programmer should not need to keep asynchrony in mind - because otherwise it's no longer "transparent".

I already wrote a number of JS programs myself in which I changed a previously synchronous approach to an asynchronous one (popular example: `localStorage` becomes `localForage`) - and actually had to apply an awful lot of changes as the move from "synchronous" to "asynchronous" changed so much...

Event handling with asynchronous handlers is especially difficult as the order of events (_any_ events, not just those of the same `type`) must be preserved - or the state machine you are implementing may no longer work as intended.

Again, if you continue to claim to be "async transparent", a script should work as intended regardless of whether certain commands are synchronous or not - the situation could even change over time - or depend on a configuration (again, you may use `localStorage` -> `localForage` as a practical example)

**Comment by 1cg** [comment]
We are talking about event bubbling here, right?  

The hyperscript scripts themselves are async transparent (as best we can do) in that, if you switch from localStorage to localForage it shouldn't require any changes to your code unless you are relying on event bubbling ordering in some way.

I would recommend reading this:

https://www.dreamsongs.com/RiseOfWorseIsBetter.html

Hyperscript isn't a formal language, it's a front end scripting language, and the goal is to be useful rather than semantically perfect.  I get where you are coming from, and I appreciate that perspective, but we have our own on the matter.  Event bubbling ordering isn't an area that is important enough to warrant additional implementation complexity, particularly given the simple work arounds available (trigger and listen for a different event).  

If it comes up a bunch in real world use cases, we can take a look at handling it, but until then I'm wary of making abstract consistency the enemy of real world simplicity.

**Comment by rozek** [comment]
No, not just event bubbling (and default actions)

We are also talking about preserving the order in which events are handled in order to keep the behaviour of state machines predictable...

**Comment by rozek** [comment]
Well,

as mentioned somewhere else: I really like the idea of

* having an xTalk language that manipulates the DOM (and has syntactic sugar for that)
* with direct JavaScript interoperability and
* "async transparency" (JS is so horrible in that aspect)

and spent some time evaluating it (e.g., to see if I could build a visual IDE around it - s.th. like _hypercard). As you may see from the number of issues I opened, there were quite a number of situations that "surprised" me and consumed a lot of time...

For the time being, I'll have to put my activities on hold for a while and watch how \_hyperscript evolves - perhaps experimenting with it again at some time in the future.

In any case, I wish you every success!

**Comment by 1cg** [comment]
Re-reading and not sure what to do besides write my own browser and VM to make this work right :)
